examples/hacspec-gimli/src/lib.rs

```diff
@@ -0,0 +1,251 @@
+use hacspec_lib::*;
+
+array!(State, 12, U32, type_for_indexes: StateIdx);
+
+fn swap(mut s: State, i: StateIdx, j: StateIdx) -> State {
+    let tmp = s[i];
+    s[i] = s[j];
+    s[j] = tmp;
+    s
+}
+
+fn gimli_round(mut s: State, r: u32) -> State {
+    for col in 0usize..4 {
+        let x = s[col].rotate_left(24);
+        let y = s[col + 4].rotate_left(9);
+        let z = s[col + 8];
+
+        s[col + 8] = x ^ (z << 1) ^ ((y & z) << 2);
+        s[col + 4] = y ^ x ^ ((x | z) << 1);
+        s[col] = z ^ y ^ ((x & y) << 3);
+    }
+
+    if (r & 3u32) == 0u32 {
+        s = swap(s, 0, 1);
+        s = swap(s, 2, 3);
+    }
+
+    if (r & 3u32) == 2u32 {
+        s = swap(s, 0, 2);
+        s = swap(s, 1, 3);
+    }
+
+    if (r & 3u32) == 0u32 {
+        s[0] = s[0] ^ (U32(0x9e377900u32) | U32(r))
+    }
+
+    s
+}
+
+pub fn gimli(mut s: State) -> State {
+    for rnd in 0..24 {
+        // XXX: Why do we only allow usize loops?
```

Comment:
we could add support for `u32`, `u8`, `u16`, etc. but then this would clutter the Hacspec typechecker code. It's also better for verification to have only one sort of loops, the proof style will be more uniform.